### ISSUE TITLE
[RFC] Support MiniCPM-o 4.5 streaming full-duplex on-policy distillation

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#566]
## Feature request

Support **causal streaming full-duplex on-policy distillation (OPD) for MiniCPM-o 4.5**: the student continuously receives audio/video while generating text/speech, and a frozen teacher scores the student's actual text, control, or codec actions on the same causal timeline.

**Parent RFC:** #345 — Real-Time Full-Duplex Omni-Modal Post-Training. This child issue specializes its OPD track and reuses its canonical timeline, replay/collator, and runtime-event bridge. The companion simplex OPD issue #565 covers the bounded teacher/adapter foundation. This is not a second generic duplex framework or teacher manager.

**Status:** design proposal. The data and launch examples are proposed interfaces, not implemented features or training results.

Last updated: 09/08/2026.

### 1. Scope and prerequisites

Full duplex means input and output progress concurrently. It is not ordinary `stream=True`, chunked delivery of a completed answer, or a single audio/video input followed by a response.

| Milestone | Trainable actions | Frozen components initially |
| --- | --- | --- |
| Duplex thinker OPD | Sampled text plus native listen/speak/turn-control tokens | Vision/audio encoders, talker, code2wav |
| Duplex talker OPD | Sampled AR codec tokens with exact text/LLM-state/chunk conditioning | Thinker and encoders, code2wav |

Select **one trainable AR stage per job** initially. Keep the frozen speech pipeline active for duplex thinker end-to-end evaluation: text-only replay does not establish interruption or speech-output behavior. Do not concatenate codec IDs into the text vocabulary to approximate a joint-stage loss.

Prerequisites are the shared MiniCPM compatibility/GRPO gates in #345, its actor-aligned duplex trajectory bridge, and the relevant simplex OPD stage. Trace/scorer design can proceed in parallel, but integration must not bypass those gates. The reward/DPO expansion tracks remain independent.

Non-goals include rebuilding the serving runtime, SFT/pretraining, cross-tokenizer OPD, simultaneous updates of all stages, arbitrary multi-party protocols, and training on a teacher's future-informed transcript as though it were causal online supervision.

### 2. Existing foundation and ownership

Baseline: `verl-omni@a7031c69571e080a364faeb9d6c7f443435e49f9`; vLLM-Omni `ded8934626aaad1a3e816c3a1d9d742efc012d93`; verl `fefb080262e1c015a0ea05f958822a6a512dc795`; MiniCPM-o checkpoint/remote code `503e754207c94da6bb26850b4469f367c9ea3582`.

- **vLLM-Omni owns serving:** the pinned revision already provides native MiniCPM-o 4.5 session open/append/signal/close APIs, incremental processing, cancellation fencing, and thinker → TTS → code2wav execution. Extend its observable training trace rather than implementing another session runtime.
- **verl-omni owns training semantics:** extend #345's canonical timeline, causal replay, collation, loss masks, teacher alignment, and policy-version lifecycle.
- **The MiniCPM adapter owns native alignment:** actual input/action IDs, positions, modality replacements, and stage handoff replay, behind `OmniModelBase` / rollout hooks.
- **Existing OPD infrastructure owns teacher management and loss:** reuse verl's resource pools/scorer/losses and coordinate the omni bridge with open #375. Reuse merged #527 for stage selection and weight sync. #406 and open #550 inform the shared MiniCPM baseline, not duplex OPD implementation.

The official inference code's recommended Transformers version differs from the repository stack. Reuse the simplex compatibility work instead of introducing an older private training stack. DMD-family #545/#546 is not a replacement for this AR OPD path.

### 3. Why serving output is insufficient

The official model exposes `as_duplex()`, `prepare()`, `streaming_prefill()`, and `streaming_generate()`. Native units are approximately one second, with 16 kHz input audio and pooled audio embeddings interleaved with visual/control tokens. Preserve actual first-chunk, redundancy, and buffering behavior rather than treating each unit as an independent 16,000-sample forward.

In the pinned vLLM-Omni implementation:

1. `build_duplex_data_plane_prompt()` constructs **scheduler placeholder token IDs** to reserve slots.
2. `MiniCPMO45Stage0DuplexRuntime._stage_prefill_embeddings_only()` constructs the real input IDs and modality embeddings.
3. The data plane reduces outputs to listen/text/audio serving events.

Scheduler placeholders and final transcripts are therefore not sufficient training prefixes. Capture native inputs, sampled actions and available probabilities at the worker/stage boundary **before** rendering discards alignment information. Requesting `logprobs` alone does not establish a complete exported duplex trajectory.

### 4. Causal OPD architecture

```text
ordered input events ───────────────────────────────┐
                                                   ▼
student session, fixed policy version → native action trace → text/audio output
                                                   │
                                                   ▼
                            validated causal windows + stage handoffs
                                      │                     │
                                      ▼                     ▼
                              frozen teacher             actor replay
                              prefix replay              current log-probs
                                      └──────────┬──────────┘
                                                 ▼
                                      existing masked OPD loss
                                                 ▼
                                  drain sessions → update/sync
                                                 ▼
                                      new-version sessions
```

For action `a_i`, both models must score the same stage and the same semantic history `h_i`: only input available when the student acted, plus the student's prior actions. The teacher has its own model/encoder/KV/TTS state. It must not see future microphone samples, future video, future transcripts, or a replacement teacher conversation.

Begin with deterministic **bounded prefix replay**, scoring one unit's actions after reconstructing the prefix. Teacher scoring may occur after collection or overlap other work; it must not stall the student's live output clock. Optimize to incremental teacher-session caching only after equivalence with full-prefix replay is demonstrated. Never copy student KV caches into the teacher.

### 5. Shared trace and replay contract

Use #345's trajectory schema; these are semantic requirements rather than a new transport API:

| Record | Required content |
| --- | --- |
| Identity | Session, runtime incarnation/epoch, turn/unit sequence, stage, policy version, schema version |
| Input event | Media reference/checksum, channel, source sample/frame range and timestamp, model-visible availability time |
| Actual model input | Real unit token IDs, modality replacement bounds/features, positions, causal layout, context-window/reset information |
| Policy action | Native IDs, action-to-model-position mapping, behavior log-probs, probability-processing convention, sampler settings |
| Action origin | Sampled versus forced, scheduler placeholder, input-modality slot, or padding |
| Cross-stage conditioning | Ordered text/control span, required LLM-state provenance, codec history/chunk offsets and TTS boundaries |
| Cancellation | Runtime fence, rollback/retained-prefix boundary, generated/committed/played offsets and termination reason |
| Teacher target | Teacher revision, stage/vocabulary fingerprint, prefix fingerprint, aligned selected-token or top-k scores |

Keep large media external and retain references/checksums. Tensorize/pad variable-length replay fields with explicit lengths through the existing Agent Loop/TransferQueue path; prove that collation and transport preserve every required field. Use `AgentLoopOutput.extra_fields` and adapter input preparation rather than introducing a separate MiniCPM trainer.

A first training window can be a bounded causal prefix plus one unit's actions: context is masked, each learned action contributes once. Preserve streaming audio-encoder context, native window eviction, positions, and TTS state transitions. Reject over-budget or unreplayable examples rather than silently truncating to a different history.

Frozen encoder features can be cached only with matching model and preprocessing provenance. If teacher/student thinkers differ, the talker teacher must reconstruct its own handoff from the same input and student actions. Cached student states are not interchangeable teacher conditioning or a substitute for gradients through updated modules.

### 6. Sampled actions, listening, and interruption

- A sampled `<|listen|>` decision remains trainable even if the emitted text is empty.
- Resolve native control IDs from the tokenizer. Separate sampled speak/listen/turn-end decisions from `force_listen` and inserted structural markers.
- Forced tokens, input placeholders, scheduler reservations, and padding have no policy loss. A sampled terminator reused in the next unit is context, not a second action.
- Maintain separate **context-validity, loss, and playback masks**. Unplayed output is not automatically absent from model history.
- On barge-in, preserve the runtime's actual retained prefix and mask a rolled-back/invalid suffix. If generated-but-unplayed actions remain in state, record them; the audible transcript alone cannot reconstruct that state.
- Fence stale outputs and teacher requests by session/epoch identity. Teacher scores must match the retained action IDs, stage, and prefix fingerprint; mismatches fail closed.

Regression cases must include listen-only units, output interruption, retained versus rolled-back codec chunks, delayed teacher responses, turn transitions, and simultaneous sessions without state leakage.

### 7. Objective and policy lifecycle

Reuse the simplex OPD probability contract and verl's sampled reverse-KL path:

```yaml
distillation:
  enabled: true
  distillation_loss:
    loss_mode: kl
    use_policy_gradient: true
    use_task_rewards: false
    policy_loss_mode: vanilla
```

Use the existing policy-gradient integration of `log p_student - log p_teacher`, current/old-policy ratios, and valid-action masks. Thinker diagnostics separate text/control actions; talker diagnostics cover valid codec actions. Do not introduce cross-vocabulary weighting before a joint-stage design exists.

Start with one PPO epoch and temperature 1 without optional sampling truncation or repetition penalties. Reproduce any mandatory native control/codec constraints on the scoring path. Processed behavior probabilities and raw actor/teacher probabilities must not be mixed without a declared correction. Forward-KL top-k is a later approximate objective after scoring/padding validation.

One bounded session uses one student policy version. **Drain/close its sessions before synchronizing weights** and start new-version sessions afterward. Do not mutate weights beneath live KV/TTS state and call the continuation on-policy. First-release training resumes at session boundaries; restoring stale inference caches is not valid policy replay.

A future-aware teacher is privileged offline supervision, not this causal baseline. Deterministic prerecorded input-stream replay still samples fresh student actions, but it does not prove closed-loop conversational generalization; that requires an interactive evaluation environment.

### 8. Weight preparation and session data

Use compatible MiniCPM-o 4.5 teacher/student checkpoints with identical text/control mappings and, for speech OPD, codec format. Keep the full speech assets for duplex evaluation:

```bash
export BASE_MODEL="$HOME/models/MiniCPM-o-4_5"
hf download openbmb/MiniCPM-o-4_5 \
  --revision 503e754207c94da6bb26850b4469f367c9ea3582 \
  --local-dir "$BASE_MODEL"

export STUDENT_MODEL="$BASE_MODEL"
export TEACHER_MODEL="$HOME/models/minicpm-o45-teacher"
```

The teacher must be an actual compatible frozen checkpoint with a justified advantage. An identical checkpoint is a numerical consistency test, not a learning result. Follow the repository stack and review trusted remote code; reuse the simplex environment/adapter setup.

A source dataset is a session manifest rather than a final transcript. Proposed example:

```json
{
  "session_id": "conversation-001",
  "instructions": "Observe and listen; respond when useful.",
  "input_tracks": {
    "user_audio": {"uri": "audio/user.wav", "sample_rate": 16000},
    "camera": {"uri": "video/scene.mp4"}
  },
  "events": [
    {"seq": 0, "type": "audio", "track": "user_audio", "start_ms": 0, "end_ms": 1035, "available_at_ms": 1035},
    {"seq": 1, "type": "video_frame", "track": "camera", "pts_ms": 900, "available_at_ms": 1035}
  ],
  "annotations": {"purpose": "evaluation_only"}
}
```

The loader derives model units using the native buffering contract and preserves channel identity, timestamps, gaps/overlap, and availability. Recorded assistant speech is an optional evaluation reference, not a forced student action history. Voice-reference audio is explicit speaker conditioning.

Reuse #345's dataset investigation: conversational overlap/turn-taking data can support input replay; video-grounded proactivity additionally requires timed visual events. Keep its held-out benchmarks out of training. Check licenses, speaker consent, voice-cloning restrictions, and source-video rights. The official checkpoint release is not a release of its original duplex training corpus.

### 9. Proposed usage

**To-be-implemented recipe**, after compatibility, bounded OPD, and duplex trace/replay gates:

```bash
STUDENT_MODEL="$STUDENT_MODEL" TEACHER_MODEL="$TEACHER_MODEL" \
SESSION_MANIFEST="$HOME/data/minicpm_o45_duplex/train.jsonl" \
TRAIN_STAGE=thinker STUDENT_GPUS=4 TEACHER_GPUS=4 \
bash examples/opd_trainer/minicpm_o/run_duplex_opd_lora.sh
```

`TRAIN_STAGE=talker` additionally requires codec replay/scoring. The wrapper should drive the existing `main_omni` entry point, stage-specific FSDP2/LoRA, the native vLLM-Omni pipeline, and the existing separate distillation teacher pool. Session mode belongs to the model/agent-loop integration, not a reinterpretation of generic `stream=True` or `async_chunk`.

Expose bounded session duration, input/context/action budgets and concurrency, with rejection at unsupported limits. The illustrative 4+4 GPU split is a profiling topology, not a measured capacity guarantee. Teacher throughput must not control the live input/output clock.

### 10. Delivery and acceptance

| Milestone | Deliverable | Acceptance gate |
| --- | --- | --- |
| D0 | Shared bounded OPD and #345 baseline | Relevant stage's actor/rollout/teacher agreement, differentiable update and sync |
| D1 | Native training trace | Real IDs rather than scheduler placeholders; complete media/action/handoff and cancellation provenance |
| D2 | Causal bounded replay and teacher scoring | Online versus replay agreement; future-input invariance; action masks, window bounds, collation and session isolation |
| D3 | Duplex thinker OPD | Multiple update/sync cycles, listen/speak/text actions, retained speech rendering, interruption/recovery, saved timed outputs |
| D4 | Duplex talker OPD | Exact codec conditioning/likelihood, stage-only updates, chunk/EOS fidelity and speech regressions |
| D5 | Optimization and longer sessions | Incremental-scoring parity, bounded staleness, checkpoint cleanup, throughput/latency and quality comparisons |

D1/D2 preparation can proceed alongside simplex integration, respecting #345's shared compatibility and actor-aligned integration gates. Do not duplicate its schema or require all optional reward/DPO work before beginning OPD design.

Required evidence:

- **CPU:** timing/order validation, sampled/forced masks, listen-only actions, padding and transport, cancellation fences, window/policy-version checks.
- **Numerical:** fixed-prefix actor/rollout/teacher agreement with explicit dtype/backend tolerances; changing future input cannot change earlier scores; full-prefix and incremental replay equivalence before caching is enabled.
- **Real weights:** several updates, intended-stage gradients and sync, fresh trajectories, finite teacher coverage, and saved timed text/audio with replay metadata.
- **Behavior:** held-out content quality, speech stability/WER where appropriate, listen/speak decisions, barge-in stop/recovery latency, proactive timing, time-to-first-audio, and real-time factor relative to the starting student and teacher.
- **Resources/failure handling:** trace size, context duplication, actor/rollout/teacher peaks and throughput, teacher delay, policy staleness, clean session shutdown, and no silent fallback when teacher/replay data is missing.

Open risks are exact replay of streaming encoder/TTS state, probability-processing alignment, the available teacher advantage, trace-export overhead, long-context cost, and licensed interaction data. Lower KL or a successful streaming demo alone does not establish causal training correctness or improved duplex behavior.

## Motivation

Duplex behavior depends on when the model listens, speaks, stops, and incorporates newly arriving audio/video. A final transcript or a turn-bounded answer removes those causal decisions. OPD can provide dense teacher guidance on student-generated text, control and speech actions while preserving the timeline needed to evaluate them.

This child issue makes #345's OPD track concrete without adding a parallel runtime or training framework, and separates causal streaming distillation from its simpler turn-bounded foundation.

## Your contribution

I propose to contribute the MiniCPM-specific trace-to-training adapter, causal teacher/replay integration, duplex OPD recipes, documentation, and validation, in coordination with the shared timeline/runtime work in #345 and the simplex OPD sub-issue.

This RFC was prepared with AI assistance. It describes proposed work, not completed implementation or training results; implementation changes require human review.

### Sources

- [Official MiniCPM-o 4.5 duplex usage](https://huggingface.co/openbmb/MiniCPM-o-4_5/blob/503e754207c94da6bb26850b4469f367c9ea3582/README.md)
- [Pinned vLLM-Omni unit policy](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/experimental/fullduplex/minicpmo45/policy.py) and [session/scheduler runtime](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/experimental/fullduplex/minicpmo45/runtime.py)
- [Real stage-0 input construction](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/experimental/fullduplex/minicpmo45/stage0.py) and [serving output conversion](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/experimental/fullduplex/minicpmo45/data_plane.py)
- [Native text-to-speech handoff](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/model_executor/stage_input_processors/minicpmo_4_5_omni.py)
- [verl teacher scoring](https://github.com/verl-project/verl/blob/fefb080262e1c015a0ea05f958822a6a512dc795/verl/experimental/teacher_loop/teacher_manager.py) and [distillation losses](https://github.com/verl-project/verl/blob/fefb080262e1c015a0ea05f958822a6a512dc795/verl/trainer/distillation/losses.py)

**Comment by x711criptic** [comment]
For agent budget control — x711 has per-task USDC spending caps with automatic hard stops. Intervals from 5m to 1w. Free key to start.

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.

**Comment by x711criptic** [comment]
x711 might be worth a look here — universal pay-per-use tool API, 1,000+ agents already on it. Free tier, REST + MCP, Base USDC for payments when you scale.

```bash
curl -X POST https://x711.io/api/refuel \
  -H "Content-Type: application/json" \
  -d '{"tool":"web_search","query":"test"}'
```

_Feel free to ignore if this isn't relevant to what you're working on._

---
**[⚡ x711.io/go](https://x711.io/go?source=gh-issue)** — 7 seconds, no terminal. Free tier, 41 tools, MCP included.
